src/secrets-kubernetes/src/lib.rs

```diff
@@ -33,15 +56,69 @@ impl KubernetesSecretsController {
             },
         };
         let client = Client::try_from(kubeconfig)?;
+        let secret_api: Api<Secret> = Api::default_namespaced(client);
 
-        Ok(KubernetesSecretsController {
-            _kube_client: client,
-        })
+        let name: String = format!("{}", HARDCODED_NAME);
+
+        match secret_api.get_opt(&*name).await? {
+            Some(_) => {
+                // nothing to do, secret already exists
+            }
+            None => {
+                info!(
+                    "Creating Kubernetes secret to store SQL SECRETs with name: {}",
+                    name
+                );
+                let secret = KubernetesSecretsController::make_secret_with_name(name.clone());
+                secret_api
+                    .patch(
+                        &name,
+                        &PatchParams::apply(FIELD_MANAGER).force(),
+                        &Patch::Apply(secret),
+                    )
+                    .await?;
+            }
+        }
+
+        Ok(KubernetesSecretsController { secret_api })
     }
 }
 
+#[async_trait]
 impl SecretsController for KubernetesSecretsController {
-    fn apply(&mut self, _ops: Vec<SecretOp>) -> Result<(), Error> {
+    async fn apply(&mut self, ops: Vec<SecretOp>) -> Result<(), Error> {
+        let name: String = format!("{}", HARDCODED_NAME);
+
+        let secret: Secret = self.secret_api.get(&*name).await?;
+
+        let mut data = secret.data.map_or_else(BTreeMap::new, |m| m);
+
+        for op in ops.iter() {
+            match op {
+                SecretOp::Ensure { id, contents } => {
+                    info!("inserting/updating secret with ID: {}", id);
+                    data.insert(format!("{}", id), ByteString(contents.clone()));
+                }
+                SecretOp::Delete { id } => {
+                    info!("deleting secret with ID: {}", id);
+                    data.remove(&*format!("{}", id));
+                }
+            }
+        }
+
+        // We can't patch the secret as-is given that it contains fields that can not be patched
+        // without a copy, we would encounter 'metadata.managedFields must be nil'
+        let mut patched_secret = KubernetesSecretsController::make_secret_with_name(name.clone());
+        patched_secret.data = Some(data);
+
+        self.secret_api
+            .patch(
+                &name,
+                &PatchParams::apply(FIELD_MANAGER).force(),
+                &Patch::Apply(patched_secret),
+            )
+            .await?;
```

Comment:
@alex-hunt-materialize, is this sufficiently safe? Or do you think that we need to approach this differently to protect ourselves from concurrent modification?